language/move-lang/src/errors/new.rs

```diff
@@ -0,0 +1,192 @@
+    for diag in diags.0 {
+        if seen.contains(&diag) {
+            continue;
+        }
+        seen.insert(diag.clone());
+        let rendered = render_diagnostic(file_mapping, diag);
+        emit(writer, &Config::default(), files, &rendered).unwrap()
+    }
+}
+
+fn convert_loc(file_mapping: &FileMapping, loc: Loc) -> (FileId, Range<usize>) {
+    let fname = loc.file();
+    let id = *file_mapping.get(fname).unwrap();
+    let start = loc.span().start().to_usize();
+    let end = loc.span().end().to_usize();
+    (id, Range { start, end })
+}
+
+fn render_diagnostic(
+    file_mapping: &FileMapping,
+    diag: Diagnostic,
+) -> csr::diagnostic::Diagnostic<FileId> {
+    use csr::diagnostic::{Label, LabelStyle};
+    let mk_lbl = |style: LabelStyle, msg: (Loc, String)| -> Label<FileId> {
+        let (id, range) = convert_loc(file_mapping, msg.0);
+        csr::diagnostic::Label::new(style, id, range).with_message(msg.1)
+    };
+    let Diagnostic {
+        info,
+        primary_label,
+        secondary_labels,
+    } = diag;
+    let mut diag = csr::diagnostic::Diagnostic::new(info.severity);
+    let (code, message) = info.render();
+    diag = diag.with_code(code);
+    diag = diag.with_message(message);
+    diag = diag.with_labels(vec![mk_lbl(LabelStyle::Primary, primary_label)]);
+    diag = diag.with_labels(
+        secondary_labels
+            .into_iter()
+            .map(|msg| mk_lbl(LabelStyle::Secondary, msg))
+            .collect(),
+    );
+    diag
+}
+
+//**************************************************************************************************
+// impls
+//**************************************************************************************************
+
+impl Diagnostics {
+    pub fn new() -> Self {
+        Self(vec![])
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn add(&mut self, diag: Diagnostic) {
+        self.0.push(diag)
+    }
+
+    pub fn extend(&mut self, diags: Self) {
+        self.0.extend(diags.0)
+    }
+
+    pub fn into_vec(self) -> Vec<Diagnostic> {
+        self.0
+    }
+}
+
+impl Diagnostic {
+    pub(crate) fn new(
+        code: impl DiagnosticCode,
+        (loc, label): (Loc, impl ToString),
+        secondary_labels: impl IntoIterator<Item = (Loc, impl ToString)>,
+    ) -> Self {
+        Diagnostic {
+            info: code.into_info(),
+            primary_label: (loc, label.to_string()),
+            secondary_labels: secondary_labels
+                .into_iter()
+                .map(|(loc, msg)| (loc, msg.to_string()))
+                .collect(),
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! diag {
+    ($code: expr, $primary: expr) => {{
+        crate::errors::new::Diagnostic::new($code, $primary, std::iter::empty::<(Loc, String)>())
+    }};
+    ($code: expr, $primary: expr,) => {{
+        diag!($code, $primary)
+    }};
```

Comment:
Nit: these two matches can be combined into one match using the `$(,)?` pattern: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f1dca056d8a2323bfe16ef6c39024590

Reply:
fancy, thanks!